src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.DiagnosticProvider.cs

```diff
@@ -42,27 +40,30 @@ public abstract class DiagnosticProvider
             /// </summary>
             public abstract Task<IEnumerable<Diagnostic>> GetAllDiagnosticsAsync(Project project, CancellationToken cancellationToken);
 
-            internal async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixAsync(
-                FixAllContext fixAllContext)
+            internal async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixAsync(FixAllContext fixAllContext)
             {
                 var result = await GetDocumentDiagnosticsToFixWorkerAsync(fixAllContext).ConfigureAwait(false);
 
                 // Filter out any documents that we don't have any diagnostics for.
                 return result.Where(kvp => !kvp.Value.IsDefaultOrEmpty).ToImmutableDictionary();
-            }
 
-            internal virtual async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixWorkerAsync(
-                FixAllContext fixAllContext)
-            {
-                using (Logger.LogBlock(
-                    FunctionId.CodeFixes_FixAllOccurrencesComputation_Document_Diagnostics,
-                    FixAllLogger.CreateCorrelationLogMessage(fixAllContext.State.CorrelationId),
-                    fixAllContext.CancellationToken))
+                static async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixWorkerAsync(FixAllContext fixAllContext)
                 {
-                    return await FixAllContextHelper.GetDocumentDiagnosticsToFixAsync(
-                        fixAllContext,
-                        fixAllContext.ProgressTracker,
-                        (document, cancellationToken) => document.IsGeneratedCode(cancellationToken)).ConfigureAwait(false);
+                    if (fixAllContext.State.DiagnosticProvider is FixAllState.FixMultipleDiagnosticProvider fixMultipleDiagnosticProvider)
+                    {
+                        return fixMultipleDiagnosticProvider.DocumentDiagnosticsMap;
+                    }
+
+                    using (Logger.LogBlock(
+                            FunctionId.CodeFixes_FixAllOccurrencesComputation_Document_Diagnostics,
+                            FixAllLogger.CreateCorrelationLogMessage(fixAllContext.State.CorrelationId),
+                            fixAllContext.CancellationToken))
+                    {
+                        return await FixAllContextHelper.GetDocumentDiagnosticsToFixAsync(
+                            fixAllContext,
+                            fixAllContext.ProgressTracker,
+                            DocumentExtensions.IsGeneratedCode).ConfigureAwait(false);
```

Comment:
❔ How do we know the new allocations will not impact performance?

Reply:
The last parameter is being removed in my follow-up PR as IsGeneratedCode and IGeneratedCodeRecognitionService will move into the shared layer. So `GetDocumentDiagnosticsToFixAsync` can directly invoke the service instead of each callsite requiring to pass this.

Reply:
it might.  i can def switch to a single-alloc form.